.github/workflows/code.yml

```
@@ -0,0 +1,191 @@
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: ghcr.io/${{ github.repository }}
+          tags: |
+            type=ref,event=branch
+            type=ref,event=tag
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Build developer image for testing
+        uses: docker/build-push-action@v3
+        with:
+          tags: build:latest
+          context: .
+          target: build
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache
+          load: true
+
+      - name: Run tests in the container locked with requirements_dev.txt
+        run: |
+          docker run --name test build bash /project/tests/container_tests.sh
+          docker cp test:/project/dist .
+          docker cp test:/project/requirements.txt .
+          docker cp test:/project/requirements_dev.txt .
+
+      - name: Build runtime image
+        uses: docker/build-push-action@v3
+        with:
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.meta.outputs.tags }}
+          context: .
+          labels: ${{ steps.meta.outputs.labels }}
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache
+
+      - name: Upload build files
+        uses: actions/upload-artifact@v3
+        with:
+          name: build_files
+          path: |
+            dist
+            requirements.txt
+            requirements_dev.txt
+
+  sdist:
+    needs: container
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - uses: actions/setup-python@v2
+        with:
+          python-version: "3.10"
+
+      - uses: actions/download-artifact@v3
+
+      - name: Install sdist in a venv and check cli works
+        # ${GITHUB_REPOSITORY##*/} is the repo name without org
+        # Replace this with the cli command if different to the repo name
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install build_files/dist/*.gz
+          ${GITHUB_REPOSITORY##*/} --version
+
+  release:
+    # upload to PyPI and make a release on every tag
+    if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
+    needs: [lint, container]
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/download-artifact@v3
+
+      - name: Github Release
+        # We pin to the SHA, not the tag, for security reasons.
+        # https://docs.github.com/en/actions/learn-github-actions/security-hardening-for-github-actions#using-third-party-actions
+        uses: softprops/action-gh-release@1e07f4398721186383de40550babbdf2b84acfc5 # v0.1.14
+        with:
+          files: |
+            build_files/*
```

CLI project has no wheel or sdist
```suggestion
            build_files/*
            build_files/dist/*
```

---

You comment and suggested change do not seem to correlate for me?

---

The CI project doesn't have a wheel or sdist attached, I'm guessing because the * isn't recursive, so the dist directory is not published, hence the other line in the suggestion

---

OK - TODO - check this works

---

it works

---

This could be simplified further by putting `requirements*.txt` in `dist` in `container_tests.sh`, then only `docker cp dist`, uploading `dist/*` as artifacts, and changing to `pipx run twine upload build_files/*.whl build_files/*.tar.gz`

---

It also emphasizes the difference between src requirements (optional and partial), and dist requirements (complete and generated)

---

I'd also be tempted to call it `dist` rather than `build_files` here too

---

simplified and switched to 'dist' as suggested.